cluster/encryption.go

```diff
@@ -0,0 +1,152 @@
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"os"
+	"strings"
+	"text/template"
+
+	"github.com/rancher/rke/k8s"
+	"github.com/rancher/rke/services"
+	"github.com/rancher/rke/templates"
+	"github.com/sirupsen/logrus"
+)
+
+// DeploySecretsEncryptionConfig -  Deploy encryption.yaml to ControlPlane hosts.
+func (c *Cluster) DeploySecretsEncryptionConfig(ctx context.Context, currentCluster *Cluster) error {
+	if currentCluster == nil {
+		currentCluster = &Cluster{}
+	}
+
+	if c.Services.KubeAPI.SecretsEncryptionConfig == currentCluster.Services.KubeAPI.SecretsEncryptionConfig {
+		logrus.Debug("[controlplane] Secrets encryption config matches state. No changes needed.")
+		return nil
+	}
+
+	for _, host := range c.ControlPlaneHosts {
+		err := c.WriteFile(ctx, host, "controlplane", "/etc/kubernetes/encryption.yaml", "0600", c.Services.KubeAPI.SecretsEncryptionConfig)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ReconcileSecretsEncryptionConfig - Restart controlplane and replace all secrets in the cluster when the encryption config changes.
+func (c *Cluster) ReconcileSecretsEncryptionConfig(ctx context.Context, currentCluster *Cluster) error {
+	if currentCluster != nil {
+		if c.Services.KubeAPI.SecretsEncryptionConfig == currentCluster.Services.KubeAPI.SecretsEncryptionConfig {
+			logrus.Debug("[reconcile] Secrets encryption config matches. No changes needed.")
+			return nil
+		}
+
+		logrus.Info("[reconcile] New secrets encryption config. Updating existing secrets to apply encryption.")
+
+		err := services.RestartControlPlane(ctx, c.ControlPlaneHosts)
+		if err != nil {
+			return err
+		}
+
+		kubeClient, err := k8s.NewClient(c.LocalKubeConfigPath, c.K8sWrapTransport)
+		if err != nil {
+			return fmt.Errorf("Failed to initialize new kubernetes client: %v", err)
+		}
+
+		secrets, err := k8s.GetAllSecrets(kubeClient)
+		if err != nil {
+			return err
+		}
+
+		for _, secret := range secrets {
+			err = k8s.UpdateSecretObject(kubeClient, &secret)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Need some logic to keep users from shooting themselves in the foot.
+// If you rotate keys without the old key in the list you will lock yourself out of the existing secrets.
+// Ways people will break this:
+// - Add a config after a cluster is up with a generated config
+// - Change a config without putting the old key in the list before re-create of secrets.
+
+// SetSecretsEncryptionConfig - Set a default encryption config and secret if not provided.
+func (c *Cluster) SetSecretsEncryptionConfig(currentCluster *Cluster) error {
+	if currentCluster == nil {
+		currentCluster = &Cluster{}
+	}
+	// cluster.yaml not defined or populated
+	if len(c.Services.KubeAPI.SecretsEncryptionConfig) == 0 {
+		// Generate from template if not populated in current config.
+		if len(currentCluster.Services.KubeAPI.SecretsEncryptionConfig) == 0 {
+			clusterMajorVersion := getTagMajorVersion(c.Version)
```

Comment:
getting the encryption config should be separated in its own method like setting the cloud config file and the result should be one of the cluster's object field that can be used later, for example https://github.com/rancher/rke/blob/master/cluster/cluster.go#L179-L181